src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -988,38 +987,36 @@ public unsafe void HandleEvents(uint events)
                     // TODO: error handling
                 }
 
-                // TODO: rethink ordering of HUP/RDHUP w.r.t. IN/OUT.
-
-                if ((events & Interop.libc.EPOLLHUP) != 0)
+                if ((events & SocketAsyncEvents.Close) != 0)
                 {
-                    // Drain queues and unregister fd
+                    // Drain queues and unregister events
 
-                    Queue<AcceptOrConnectOperation> acceptOrConnectQueue;
-                    Queue<TransferOperation> sendQueue;
-                    Queue<TransferOperation> receiveQueue;
+                    OperationQueue<AcceptOrConnectOperation> acceptOrConnectQueue;
+                    OperationQueue<TransferOperation> sendQueue;
+                    OperationQueue<TransferOperation> receiveQueue;
                     lock (_queueLock)
                     {
                         acceptOrConnectQueue = _acceptOrConnectQueue.Stop();
                         sendQueue = _sendQueue.Stop();
                         receiveQueue = _receiveQueue.Stop();
 
-                        if (_registeredEvents != 0)
+                        if (_registeredEvents != SocketAsyncEvents.None)
                         {
                             Interop.Error errorCode;
-                            if (!_engine.TryUnregister(ref _handle, _fileDescriptor, 0, out errorCode))
+                            if (!_engine.TryRegister(_fileDescriptor, _registeredEvents, SocketAsyncEvents.None, _handle, out errorCode))
```

Comment:
Nevermind, I see now that TryRegister handles both add and delete based on the args supplied to it.